### DISCUSSION TITLE
Inconsistent loss between the predict method and the 'loss' column in the equations_ dataframe

### DISCUSSION THREAD

**Discussion by lerouxerwan** [MilesCranmer/PySR#943] [Q&A] [answered]
### What happened?

While doing some tests on a toy dataset, I may have noted some inconsistency for the train loss between the predict method and the 'loss' column in the equations_ dataframe. 

These inconsistencies, i.e. small differences, cannot always be explained with rounding. In practice, such inconsistencies can have undesirable consequences. 

For instance, for the toy dataset/example below, some equations that are in the Pareto optimal set of equations (they are in the equations_ dataframe, where the loss is indeed decreasing with complexity) should not be in this set of equations if we consider the loss from the predict method.

Example, in the log output below, at the index 3 and 4:
-the loss from equations_ dataframe are 0.8148807 and 0.8148479 (0.8148807 > 0.8148479)
-the loss from the predict method are 0.8149047310175083 and 0.8149065628195199  (0.8149047310175083 < 0.8149065628195199)

### Version

1.5.8

### Operating System

Linux

### Package Manager

pip

### Interface

Script (i.e., `python my_script.py`)

### Relevant log output

```shell
Loss from equations dataframe: [18581504.0, 12565.13, 1070.6243, 0.8148807, 0.8148479, 0.81479234, 0.81172127, 0.7941829] 

Loss from predict method: [18581503.45889028, 12565.122104296868, 1070.6286069439932, 0.8149047310175083, 0.8149065628195199, 0.8149049940477293, 0.8118029493713993, 0.7940984955826653]
```

### Extra Info

```python
import numpy as np
from pysr import PySRRegressor
from scipy.stats import norm
from sklearn.metrics import mean_squared_error

def main():

    # Load toy dataset, containing 100 datapoints with one feature and one target
    size = 100
    X = np.expand_dims(np.arange(size), axis=-1).astype(float)
    y = X[:, 0] ** 2 - 2 * X[:, 0] + 3 + norm.rvs(loc=0, scale=1, size=size, random_state=42)

    # Fit model
    model = PySRRegressor(niterations=1, verbosity=0)
    model.fit(X, y)

    # Show loss from the "equations_" dataframe
    loss_values_from_equations_dataframe = [float(value) for value in model.equations_['loss'].values]
    print('Loss from equations dataframe:', loss_values_from_equations_dataframe, '\n')

    # Show loss found using the predict method
    y_predict_list = [model.predict(X, index=index) for index in range(len(model.equations_))]
    loss_values_from_predict_method = [mean_squared_error(y, y_predict) for y_predict in y_predict_list]
    print('Loss from predict method:', loss_values_from_predict_method)

if __name__ == '__main__':
    main()
```

**Comment by MilesCranmer** [comment]
This could just be 32-bit vs 64-bit evaluation? Try to set precision=64 in the PySRRegressor

**Comment by lerouxerwan** [comment]
Thanks for your quick answer, I modified the code with precision=64, and it gives:

Loss from equations dataframe: [18555242.857113197, 12565.122104296866, 931.1322485436953, 0.8149047134979738, 0.814904713497913, 0.8117986875470037, 0.811798687546914, 0.8002984905737854] 

Loss from predict method: [18555242.857113194, 12565.122104296868, 931.1322485436951, 0.8149047134979738, 0.8149047134979434, 0.8117986875469806, 0.8117986875470672, 0.8002984905737895] 

In term of differences (loss from equations_ dataframe MINUS loss from predict method) it gives:

Difference between the two loss values: [3.725290298461914e-09, -1.8189894035458565e-12, 1.1368683772161603e-13, 0.0, -3.0309088572266774e-14, 2.3092638912203256e-14, -1.532107773982716e-13, -4.107825191113079e-15]

The difference is equal to zero only for one equation. Otherwise, the difference is extremely small. 

However, in this toy example, despite these small differences, the issue mentioned above can persist (the fact that an equation initially in the Pareto-optimal set with the loss from equations_ dataframe, is not in the Pareto-optimal set if we consider the loss from the predict method)

For instance, at index 6 and 7 :

-the loss from equations_ dataframe are 0.81179868754**70037** > 0.81179868754**6914**

-the loss from the predict method are  0.81179868754**69806** < 0.81179868754**70672**

To conclude, differences with precision=64 are extremely small so it is probably not a big issue (except if you would have expected exactly the same values with 64-bit), however it is more disturbing the fact that an equation can be Pareto-optimal if the loss is taken from the dataframe and not Pareto-optimal if the loss is recomputed with the predict method.

**Comment by lerouxerwan** [comment]
To circumvent this issue, and be sure that every equations considered are Pareto-optimal with respect to the loss from the predict method, for the moment in the rest of my code I :

1) fit with PySRRegressor

2) recompute the ‘loss’ column of the equations_ dataframe using the predict method

3) eliminate any row/equation from the dataframe that is not Pareto-optimal anymore

**Comment by MilesCranmer** [comment, marked as answer]
I think it is half about the precision, and then half about the order in which the expressions are evaluated. I don't think there is much to do about this though. 

This sort of difference is actually a fundamental limitation of floating point numerical calculations! To see this, open up Python and evaluate the two expressions below:

```python
[ins] In [1]: 0.1 + (0.2 - 0.3)
Out[1]: 2.7755575615628914e-17

[ins] In [2]: (0.1 + 0.2) - 0.3
Out[2]: 5.551115123125783e-17
```

Indeed, even though these are supposed to both be 0, they actually evaluate to slightly different things depending on how you evaluate them! Does that mean one of these expressions is "greater in value" than the other? Not necessarily. When you say something is greater than something else, you need to also say how you are evaluating it.

In other words, it is not as though sympy evaluation + `sklearn.metrics.mean_squared_error` compared to the Julia equivalent (used for generating the table) is more correct than the other. They make slightly different choices in how they evaluate these things.

In this sense, you should not think of some sort of absolute "Pareto optimal" ordering, but, rather, "Pareto optimal **according to the evaluation scheme and dataset**."  If you change the evaluation scheme, or change the dataset, it is not unreasonable to get slightly different Pareto optimality.

---

If it is very important for your application that you have an absolute ordering down to infinite precision, you could consider using rational numbers for evaluation (with Julia's `BigInt`). This needs to be done on the Julia side though; see https://ai.damtp.cam.ac.uk/symbolicregression/dev/examples/custom_types/. However, this is going to be prohibitively slow. So it's probably worth just living with the weirdness of floating point numbers.

To see how rational numbers avoid this issue:

```julia
julia> 1//10 + (2//10 - 3//10)
0//1

julia> (1//10 + 2//10) - 3//10
0//1
```

**Comment by lerouxerwan** [comment]
Thanks a lot for this thorough explanation

It's been quite a long time since I learnt about floating point evaluation, thanks for all this reminder & the simple example, it's always fascinating to remind ourselves that ordering of operations matter with such evaluation.

Initially (with 32-bits precision) the difference between the 2 losses were larger, which is why I posted this issue (because I thought that an issue might have gone unnoticed between the Julia code & the Python code).

The idea of an absolute ordering is great. For simplicity, I think I will stick to my current methodology described above (rely on the loss with the “predict method” so that all my losses remain consistent across all my results/code, which is quite important for some ‘model selection criterion’)
 
Thanks again,

Erwan
